### PR TITLE
[WIP] Feat: Performance Test Auxiliaries

### PR DESCRIPTION
**Pull request by Somefive**
**What this PR does / why we need it**:

This PR refactors loggings in controller by aggregating functions in ReconcileContext. The new ReconcileContext has the following benefits:
1. It can easily add checkpoints to track time costs.
2. It normalizes the logging format in controllers which enables effective logging collection by Loki in monitoring system.
3. Later controller-level modifications to the context can be easily extended to all controllers.

In addition, several files are added to the `./hack/pressure-test` which provides a simple guide for running KubeVela performance test.

### CODE REVIEW

File: hack/pressure-test/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Pressure Test Guide
+
+## Files
+
+Two application templates are provided in `app-templates` folder. The community template is a standard template which can be applied in standard KubeVela environment. The sae template, on the other hand, requires support for extended definitions (env & cloneset-service) and OpenKruise to be installed. The `tolerate-hollow-node` trait is used to add toleration to pods in pressure test applications, which makes them allocatable for hollow nodes mocked by `kubemark`.
```
Comment (zzxwill) [inline hack/pressure-test/README.md:5]:
```suggestion
Two application templates are provided in [app-templates](./app-templates) folder. The community template is a standard template which can be applied in standard KubeVela environment. The sae template, on the other hand, requires support for extended definitions (env & cloneset-service) and OpenKruise to be installed. The `tolerate-hollow-node` trait is used to add toleration to pods in pressure test applications, which makes them allocatable for hollow nodes mocked by `kubemark`.
```

File: hack/pressure-test/README.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+The `dashboard.json` is a grafana dashboard file that can track several groups of key metrics when running pressure tests.
+
+## Steps
+
+1. **Prepare Kubernetes Clusters.** The pressure test is conducted under ACK (Alibaba Cloud Container Service for Kubernetes) environment. You can set up your own Kubernetes cluster environment if you would like to run your own pressure test. 
+2. **Prepare Kubernetes Environments.** You need to install KubeVela first in your environment and add environment variable `PERF=1` to the pods of KubeVela to enable performance monitoring. Remember to disable MutatingWebhookConfigurations and ValidatingWebhookConfigurations if you only care about the KubeVela controller performance.
```
Comment (zzxwill) [inline hack/pressure-test/README.md:12]: You can add a link to specify how to add the environment https://github.com/oam-dev/kubevela/blob/master/charts/vela-core/templates/kubevela-controller.yaml

File: hack/pressure-test/README.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+3. **[Optional] Prepare Kubemark Environments.** The `tolerate-hollow-node` trait need to be installed if you would like to make pressure test on hollow nodes mocked by `kubemark`. Another issue for using `kubemark` is that you need to ensure pods created by pressure test applications would be assigned to hollow nodes while other pods like KubeVela controller must not be assigned on these hollow nodes. You can achieve that by adding taints to nodes and adding nodeSelector or toleration to pods.  
+4. **Prepare Monitoring Tools.** Install Grafana, Loki, Prometheus in your Kubernetes which can help you watch pressure test progress and discover problems.
```
Comment (zzxwill) [inline hack/pressure-test/README.md:14]: Use `observability` addon?

**Comment by zzxwill** [comment]
Will `klog` be replaced by `ctx`?

File: hack/pressure-test/README.md (hunk 3 above)
Comment (Somefive) [reply hack/pressure-test/README.md:14]: > Use `observability` addon?

Ok, I added it. Using addon will require user to install `vela` first. The usage of it is introduced as an optional choice.

**Comment by Somefive** [comment]
> Will `klog` be replaced by `ctx`?

Yes, there are several advantages of using `ctx` which wraps some extra codes on top of `klog`, as mentioned in the PR introduction.

1. `ReconcileContext` wraps the target object namespace and name in its key/value pair which makes all logs in controller carry this information.
2. `ReconcileContext` gives 1 minute time limit for each reconcile.
3. `ReconcileContext` automatically extracts controller file location and extract the controller name which is carried for each log as well.
4. `ReconcileContext` outputs all logs with key/values described above in fixed format which can be easily collected by logging services such as Loki.
5. `ReconcileContext` records time cost for each reconcile by default and can track details optionally.

File: hack/pressure-test/README.md (hunk 2 above)
Comment (Somefive) [reply hack/pressure-test/README.md:12]: I changed it to command line flags as recommended by @wonderflow . Thanks for the review.

File: pkg/controller/core.oam.dev/v1alpha2/application/revision_test.go [hunk 4]
```diff
@@ -466,7 +470,7 @@ var _ = Describe("test generate revision ", func() {
 	It("Test App with rollout template", func() {
 		By("Apply the application")
 		appParser := appfile.NewApplicationParser(reconciler.Client, reconciler.dm, reconciler.pd)
-		ctx = util.SetNamespaceInCtx(ctx, app.Namespace)
+		// ctx = util.SetNamespaceInCtx(ctx, app.Namespace)
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/revision_test.go:473]: why deleted

Comment (wonderflow) [reply pkg/controller/core.oam.dev/v1alpha2/application/revision_test.go:473]: several same questions below

File: pkg/controller/core.oam.dev/v1alpha2/applicationrollout/helper.go [hunk 5]
```diff
@@ -17,10 +17,15 @@ limitations under the License.
...
 import (
-	"context"
 	"fmt"
 	"reflect"
 
+	"github.com/pkg/errors"
+
+	"github.com/oam-dev/kubevela/pkg/controller/common"
+
+	"github.com/go-logr/logr"
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/applicationrollout/helper.go:27]: format the group

**Review by wonderflow — APPROVED** [review]
general LGTM， ping @leejanee @captainroy-hy  please review

**Comment by leejanee** [comment]
suggest to make reconcileContext to have the ability to fork

File: pkg/controller/core.oam.dev/v1alpha2/application/apply.go [hunk 6]
```diff
@@ -57,14 +59,19 @@ type AppHandler struct {
 // Dispatch apply manifests into k8s.
 func (h *AppHandler) Dispatch(ctx context.Context, manifests ...*unstructured.Unstructured) error {
 	h.initDispatcher()
-	_, err := h.dispatcher.Dispatch(ctx, manifests)
+	rctx := common2.NewReconcileContext(ctx, types2.NamespacedName{
```
Comment (captainroy-hy) [inline pkg/controller/core.oam.dev/v1alpha2/application/apply.go:62]: Why create a new ctx for Dispatch() but re-use ReconcileContext passed in DispatchAndGC() ?

File: pkg/webhook/core.oam.dev/v1alpha2/application/validating_handler.go [hunk 7]
```diff
@@ -68,27 +69,33 @@ func (h *ValidatingHandler) InjectDecoder(d *admission.Decoder) error {
 }
 
 // Handle validate Application Spec here
-func (h *ValidatingHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+func (h *ValidatingHandler) Handle(_ctx context.Context, req admission.Request) admission.Response {
+	ctx := common.NewReconcileContext(_ctx, types.NamespacedName{Name: req.Name, Namespace: req.Namespace})
```
Comment (captainroy-hy) [inline pkg/webhook/core.oam.dev/v1alpha2/application/validating_handler.go:73]: Why validating_handler use new reconcile context, but mutating_handler below still use klog?

File: cmd/core/main.go [hunk 8]
```diff
@@ -103,17 +105,21 @@ func main() {
...
+	flag.BoolVar(&commonconfig.PerfEnabled, "perf-enabled", false, "Enable performance logging for controllers, disabled by default.")
 
 	flag.Parse()
 	// setup logging
 	klog.InitFlags(nil)
 	if logDebug {
-		_ = flag.Set("v", strconv.Itoa(int(commonconfig.LogDebug)))
+		_ = flag.Set("v", strconv.Itoa(commonconfig.LogDebug))
 	}
+	klog.InfoS("configs", "syncPeriod", syncPeriod.Seconds(), "qps", qps, "burst", burst, "concurrent-reconciles", controllerArgs.ConcurrentReconciles)
```
Comment (captainroy-hy) [inline cmd/core/main.go:122]: If this is the only one place where we use klog, and in other places, it just uses `klog.KObj()`, why just remove all klog in kubevela. Two styles of logger may result in confusion and abusing.

File: pkg/controller/core.oam.dev/v1alpha2/application/apply.go (hunk 6 above)
Comment (Somefive) [reply pkg/controller/core.oam.dev/v1alpha2/application/apply.go:62]: A type check should be there. I will fix it.

File: cmd/core/main.go (hunk 8 above)
Comment (Somefive) [reply cmd/core/main.go:122]: Ok, I will fix it for other handlers.

File: pkg/webhook/core.oam.dev/v1alpha2/application/validating_handler.go (hunk 7 above)
Comment (Somefive) [reply pkg/webhook/core.oam.dev/v1alpha2/application/validating_handler.go:73]: I will change all of them. It is a bit complex.

**Comment by Somefive** [comment]
This PR is too large. I will try to split some.

**Pull request closed without merging**